cmake: Handle MPU_ALIGN in CONFIG_CMAKE_LINKER_GENERATOR

bjorniuppsala opened the pull request:
Until now it has not been possible to handle MPU_ALIGN(region_size) entierly inside the CMAKE_LINKER_GENERATOR machinery, since it has been relying on the c-preprocessor definition. This does not scale for e.g. the iar-linker, which would require a different include-file duplicating the MPU_ALIGN definition. It also makes the logic hard to follow, since the info is spread out between the cmake-files, #include files, all brought guarded with #ifdefs. Doing it all in cmake allows us to collect all this at the same place.

This patch tries to fix that by extending the already existing @variables@ to handle arguments using the syntax @MPU_ALIGN,region_size=42@ which sets @region_size@ to 42 while expanding @MPU_ALIGN@. The syntax is choosen to not interfere with cmake...

bjorniuppsala committed:
cmake: linker generator: Add recursive @variables@

To allow funcitonallity like #define MPU_ALIGN(region_size) we
need some more power from @variables@. This adds two things:

* Recursive expansion of variables, if V1=foo and V2=@V1@ then
  @V2@ will expand to foo
* @WITH_ARG:arg=foo@ defines arg=foo during the expansion of
  WITH_ARG.
* @SOME:undef=baz@ will expand to baz if SOME is undefined
* Allow @(<expression>)@ to allow proper nesting.

Together these two features allows one to do things similar to
preprocessor macros with arguments, e.g. MPU_ALIGN(reg_size):
zephyr_linker_include_var(VAR MPU_ALIGN VALUE "1 << LOG2CEIL(@rs@)")
zephyr_linker_section(NAME foo ALIGN @MPU_ALIGN:rs=foo_end-foo_start@)

Signed-off-by: Björn Bergman <bjorn.bergman@iar.com>

bjorniuppsala committed:
cmake: linker generator: MPU_ALIGN for CONFIG_CMAKE_LINKER_GENERATOR

To work around different linkers expression power, e.g. LOG2CEIL() or
align blocks to a function of their own size, this adds machinery for
tagging @variables@ into @Z_EVALUATE:undef=X:expr=....@

Those are picked up by linker_script_common.cmake and placed in a
json-file for later processing. After linking gen_z_evaluated.py picks up
the json file and does the calculations (using the prevously built elf to
lookup symbol values), and generates @Z_EVALUATED_....@ variable
definitions that can be picked up in the next linking pass.

Implement MPU_ALIGN for CONFIG_CMAKE_LINKER_GENERATOR for cortex-m, and
setup SMEM_PARTITION_ALIGN to use MPU_ALIGN. Following the pattern of
how it is setup in the non-generator scrips.

Make gen_app_partitions.py pass size argument to SMEM_PARTIOTION_ALIGN
and fix a trivial bug with both ALIGN(x) and ALIGN_WITH_INPUT in
gen_app_partitions.py

Signed-off-by: Björn Bergman <bjorn.bergman@iar.com>

tejlmand commented on cmake/linker_script/arm/linker.cmake:
```diff
@@ -22,16 +22,15 @@ if((NOT DEFINED CONFIG_CUSTOM_SECTION_ALIGN) AND DEFINED  CONFIG_MPU_REQUIRES_PO
   # Handling this requires us to handle log2ceil() in iar linker since the size
   # isn't known until then.
-  set(MPU_ALIGN_BYTES ${region_min_align})
+  zephyr_linker_include_var(VAR MPU_ALIGN VALUE "MAX(${region_min_align} , 1 << LOG2CEIL(@region_size@) )")
```
`LOG2CEIL` is problematic because it's not universally supported across linkers.

bjorniuppsala replied:
Have you looked at my IAR ILink solution ( https://github.com/zephyrproject-rtos/zephyr/commit/4ceaa1589820802aa8b249c3e5ad55a194644d1c ) ? 

It expands e.g. LOG2CEIL with a special tag (@IAR_EVALUATE@ ), which the generator at each linking pass puts into a json list of required expressions. This becomes an extra output from the generator, and after linking the pass a ruby script looks at the list of expressions and evaluates them, using the linking result. In the next linking pass the values can be used, at least if not too much has changed (which it hasn't between pre1 and final, for MPU_ALIGN). 

This is reasonably well hidden behind @variables@ evaluation, so the impact on the generator itself is acceptable I think.
The python script it self needs some work to be "industrial strength", but I think the overall setup works.

It is rather cumbersome though, and quite hard to follow, and not what you usually expect to happen when "just linking" a project. But I see no other alternatives really, besides removing the problem altogether and disallow log2ceil for all and just force the user to set a kcnofig variable with the approriate alignment manually. It is easy enoguh to have a  link time check that it follows the requirements.

bjorniuppsala replied:
I am heading out on vacation the next two weeks (until June 30th). I might check in here, but please be patient ;)

I would very much like input or suggestions on how to do this, and bridge the gap between linkers.
When I did this I thought it would be only ILink that needed the extra help with log2ceil and such...

tejlmand replied:
> Have you looked at my IAR ILink solution ( [4ceaa15](https://github.com/zephyrproject-rtos/zephyr/commit/4ceaa1589820802aa8b249c3e5ad55a194644d1c) ) ?

yes I did, and especially I find constructs like this problematic:
https://github.com/zephyrproject-rtos/zephyr/blob/4ceaa1589820802aa8b249c3e5ad55a194644d1c/cmake/linker_script/arm/linker.cmake#L26-L33

because if we allow ld functions to be written directly then we cannot expect an author to know all possible linkers out there, not to mention it becomes hard to have an out-of-tree linker generator implementation.

Also it allow code to enter and fail for other linkers, for example for IAR until @bjorniuppsala  becomes aware of the extra case.

You may say that relying on a function also requires for someone writing the functionality, like in case above first author will likely add the GNU ld support. An oot linker can implement their own.

And you can implement the IAR variant when you have time.
Also it avoids this `if` this toolchain testing everywhere, cause the function is just called, and toolchain provides the implementation (and with failure / unsupported error message if not ready yet.).

bjorniuppsala committed:
cmake: Refactor linker pass boilerplate

Some of the boilerplate when setting up the linker passes has been
repeated with very slight variations. This moves it into a macro
(since a big part is calling the linker-specific
configure_linker_script, which is a macro)

Signed-off-by: Björn Bergman <bjorn.bergman@iar.com>
